src/test/java/com/trilead/ssh2/KnownHostsTest.java

```diff
@@ -101,4 +132,52 @@ public void initializeInvalidHostKey() throws IOException {
 		obj.addHostkeys("also not-a-host-key".toCharArray());
 		assertThat(obj.publicKeys.isEmpty(), is(true));
 	}
+
+	@Test
+	public void getPreferredServerHostkeyAlgorithmOrder_RsaKey() throws Exception {
+		KnownHosts obj = new KnownHosts();
+		obj.addHostkeys(getKnownHosts("known_hosts"));
+		assertThat(obj.getPreferredServerHostkeyAlgorithmOrder("rsa"),
+			arrayContaining(
+				equalTo("rsa-sha2-512"),
+				equalTo("rsa-sha2-256"),
+				equalTo("ssh-rsa"),
+				equalTo("ssh-ed25519"),
+				equalTo("ecdsa-sha2-nistp256"),
+				equalTo("ecdsa-sha2-nistp384"),
+				equalTo("ecdsa-sha2-nistp521"),
+				equalTo("ssh-dss")));
+	}
+
+	@Test
+	public void getPreferredServerHostkeyAlgorithmOrder_EcdsaP256Key() throws Exception {
+		KnownHosts obj = new KnownHosts();
+		obj.addHostkeys(getKnownHosts("known_hosts"));
+		assertThat(obj.getPreferredServerHostkeyAlgorithmOrder("ecdsap256"),
+			arrayContaining(
+				equalTo("ecdsa-sha2-nistp256"),
+				equalTo("ssh-ed25519"),
+				equalTo("ecdsa-sha2-nistp384"),
+				equalTo("ecdsa-sha2-nistp521"),
+				equalTo("rsa-sha2-512"),
+				equalTo("rsa-sha2-256"),
+				equalTo("ssh-rsa"),
+				equalTo("ssh-dss")));
+	}
+
+	@Test
+	public void getPreferredServerHostkeyAlgorithmOrder_Multiple() throws Exception {
+		KnownHosts obj = new KnownHosts();
+		obj.addHostkeys(getKnownHosts("known_hosts"));
+		assertThat(obj.getPreferredServerHostkeyAlgorithmOrder("multiple"),
+			nullValue());
```

Comment:
i'm not sure why this should return null?

if you can't set preferred hostkey algorithms, and you have e.g. rsa and ecdsa keys, and the server has those and also ed25519 which it prefers, the verifier will receive ed25519 and fail?

Reply:
Yeah, that's why I thought this code was strange. Maybe it should just look at what keys it has, sort those according to the preference order, then append the key types it doesn't have in the remaining preference order. Otherwise it returns the globally preferred key type.

Reply:
as i understand it, `getPreferredServerHostkeyAlgorithmOrder` is supposed to be used with `Connection.setServerHostKeyAlgorithms`. so i think that the idea here is to return only those algorithms that `KnownHosts` can actually use with the server. as it's reasonable for `KnownHosts` to have several algorithms of the same or different type for any given host, i think it should return all of those in some order that could be defined by the order of the keys in the source file (original..ish behavior?) or perhaps by some global list of preferred algorithms.

for instance, if `KnownHosts` has an RSA and Ed25519 keys for `example.com`, it could return something along `["rsa-sha2-512", "ssh-ed25519", "rsa-sha2-256"]` if it considered ssh-ed25519 worse than rsa-sha2-512 but better than rsa-sha2-256.

i don't see why would `KnownHosts` return other types of keys if those keys can't be used to verify the server.

just so that i'm clear on this `null` sutuation: i had a known hosts file on my client device which had RSA and ECDSA host keys and i added Ed25519 key to the server. i could not connect, then. i put a logging statement on `ServerHostKeyVerifier.verifyServerHostKey` and saw that it's getting called only once with `Ed25519` algorithm. i take it the server was preferring that. 

i could use `ExtendedServerHostKeyVerifier.getKnownKeyAlgorithmsForHost` but this basically requires the same information as i think should be put into `setServerHostKeyAlgorithms`. the only method that can provide this information would be `getPreferredServerHostkeyAlgorithmOrder`, and it was returning `null`. it was returning null as i had two keys of different type for the same host. so maybe to marry `getKnownKeyAlgorithmsForHost` and `KnownHosts` there could be another method?..